swim/self_evict.go

```diff
@@ -0,0 +1,244 @@
+package swim
+
+import (
+	"errors"
+	"math"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/ringpop-go/util"
+
+	"github.com/uber-common/bark"
+)
+
+var (
+	// ErrDuplicateHook is returned when a hook with the given name has already
+	// been registered.
+	ErrDuplicateHook = errors.New("expected hook name to be unique")
+
+	// ErrEvictionInProgress is returned when ringpop is already in the process
+	// of evicting
+	ErrEvictionInProgress = errors.New("ringpop is already executing an eviction")
+)
+
+// SelfEvict defines the functions that interact with the self eviction of nodes
+// from the membership prior to shutting down
+type SelfEvict interface {
+	RegisterSelfEvictHook(hooks SelfEvictHook) error
+	SelfEvict() error
+}
+
+// SelfEvictHook is an interface describing a module that can be registered to
+// the self eviction hooks
+type SelfEvictHook interface {
+	// Name returns the name under which the eviction should be regitered
+	Name() string
```

Comment:
@mennopruijssers originally implemented it with a name to prevent a system like `sevnup` or `serial` to register twice. By giving them a name we can prevent such cases. Besides that it can be used to add more specific information to log messages.

That said, I would also be happy to model it more after the `EventEmitter` where hooks are registered and deregistered by identity. This would also prevent the hooks from being registered twice, however we will lose the opportunity to add the name to the logs.